cli/src/cmd_instance.rs

```diff
@@ -282,3 +282,161 @@ impl RunnableCmd for CmdInstanceFromImage {
+            .instance_external_ip_attach()
+            .instance(&id.instance);
+        if let Some(project) = &id.project {
+            builder = builder.project(project)
+        }
+
+        let result = builder.body(self).send().await;
+
+        match result {
+            Ok(r) => {
+                println!("success\n{r:#?}")
+            }
+            Err(r) => {
+                println!("error\n{r:#?}")
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Detach an external IP from an instance
+#[derive(Parser, Debug, Clone)]
+#[command(verbatim_doc_comment)]
+#[command(name = "detach")]
+pub enum CmdInstanceExternalIpDetach {
+    /// Detach and deallocate the ephemeral IP from an instance
+    Ephemeral {
+        #[clap(flatten)]
+        instance: InstanceIdentity,
+    },
+    /// Detach a floating IP from an instance
+    Floating {
+        #[clap(flatten)]
+        instance: InstanceIdentity,
+
+        /// The name of an existing floating IP within the target
+        /// instance's project.
+        #[clap(long, short)]
+        floating_ip: Name,
+    },
+}
+
+impl From<&CmdInstanceExternalIpDetach> for ExternalIpDelete {
+    fn from(value: &CmdInstanceExternalIpDetach) -> Self {
+        match value {
+            CmdInstanceExternalIpDetach::Ephemeral { .. } => Self::Ephemeral,
+            CmdInstanceExternalIpDetach::Floating { floating_ip, .. } => {
+                Self::Floating(floating_ip.clone())
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl RunnableCmd for CmdInstanceExternalIpDetach {
+    async fn run(&self, ctx: &crate::context::Context) -> Result<()> {
+        let id = match self {
+            Self::Ephemeral { instance, .. } => instance,
+            Self::Floating { instance, .. } => instance,
+        };
+
+        let mut builder = ctx
+            .client()
+            .instance_external_ip_detach()
+            .instance(&id.instance);
+        if let Some(project) = &id.project {
+            builder = builder.project(project)
+        }
+
+        let result = builder.body(self).send().await;
+
+        match result {
+            Ok(r) => {
+                println!("success\n{r:#?}")
+            }
+            Err(ref e @ oxide::Error::InvalidResponsePayload(ref r)) if r.is_decode() => {
+                // This is a minor hack to account for the case where a double
+                // ephemeral detach returns null -- the operation was a success,
+                // for idempotency, but we no longer know what the actual IP object
+                // referenced is/would be.
+                let formatted = format!("{r}");
+                if formatted.contains("invalid type: null, expected struct ExternalIp") {
+                    println!("success\n()")
+                } else {
+                    println!("error\n{e:#?}")
+                }
+            }
+            Err(r) => {
+                println!("error\n{r:#?}")
+            }
+        }
```

Comment:
I feel like I'm overlooking a way on the omicron side to avoid having to do this -- the return type there is `Option`'d, but this is not reflected in the output OpenAPI spec.

Reply:
You're missing something but it's not at all obvious. Top-level `Option` around a response does not work well. The good way to do this is to put it inside a non-optional struct:

```rs
struct MyResponseBody {
  key: Option<Value>
}
```

https://github.com/oxidecomputer/omicron/pull/4694/files#diff-7429b402f97c8a901b88c3599e8505de4afae78092b5f27e4542bc5818125f17R3779

Reply:
As noted in omicron, I think the answer is likely to respond with an error if the entity we're trying to detach is not found.